### PR TITLE
services/regulated-assets-approval-server/internal/serve:  Handle success response.

### PR DESCRIPTION
**Pull request by Zagan202**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

1.  Receive incoming transaction
2.  Inspect if transactionthe transaction was found compliant and signed(if not signed then have the issuer sign it) without being revised.
3. Return a success response
```json
{
  "status": "success",
  "tx": "AAAAAHAHhQtYBh5F2zB7...",
  "message": "Transaction is compliant and signed by the issuer, ready for the source account's signature and submission."
}
```

Closes:#https://github.com/stellar/nebula/issues/189

### Why

Covers the [SEP-08 "success" case](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0008.md#success)

### Known limitations

N/A

### CODE REVIEW

**Comment by marcelosalloum** [comment]
> 2. Inspect if transaction is a revised compliant and signed by the issuer(if not signed then have the issuer sign it).

I don't think this phrase is accurate, the word "revised" means something that was changed/edited. In this case the incoming transaction is already compliant with the regulated assets authorization strategy described in SEP-8, meaning no revision took place. Does that make sense?

**Comment by marcelosalloum** [comment]
Despite checking the "PR checklist" option _"This PR's title starts with name of package that is most changed in the PR"_, this PR title is not starting with the full package name, which should be `services/regulated-assets-approval-server/internal/serve`

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go [hunk 1]
```diff
@@ -133,6 +133,205 @@ func TestConvertThresholdToReadableString(t *testing.T) {
...
+				},
+			},
+			BaseFee:    txnbuild.MinBaseFee,
+			Timebounds: txnbuild.NewInfiniteTimeout(),
+		},
+	)
+	require.NoError(t, err)
+
+	// TEST "rejected" response if transaction source account seq num is not equal to account sequence+1.
+	require.NoError(t, err)
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:318]: This error assertion can be deleted, right?

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go [hunk 2]
```diff
@@ -564,3 +763,311 @@ func TestTxApproveHandlerTxApprove(t *testing.T) {
...
+		On("AccountDetail", horizonclient.AccountRequest{AccountID: issuerAccKeyPair.Address()}).
+		Return(horizon.Account{
+			AccountID: issuerAccKeyPair.Address(),
+			Sequence:  "1",
+			Balances: []horizon.Balance{
+				{
+					Asset:   base.Asset{Code: "ASSET", Issuer: issuerAccKeyPair.Address()},
+					Balance: "0",
+				},
+			},
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:793]: ❓ Why did you add this balance here?

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go [hunk 3]
```diff
@@ -564,3 +763,311 @@ func TestTxApproveHandlerTxApprove(t *testing.T) {
...
+	_, err = handler.db.ExecContext(ctx, updateAccountKycQuery, "xEmail@test.com", senderAccKP.Address())
+	require.NoError(t, err)
+	rejectedResponse, err = handler.checkIfCompliantTransaction(ctx, kycReqCompliantTx)
+	require.NoError(t, err)
+	wantRejectedResponse = txApprovalResponse{
+		Status:     "rejected",
+		Error:      "Your KYC was rejected and you're not authorized for operations above 500.00 GOAT.",
+		StatusCode: http.StatusBadRequest,
+	}
+	assert.Equal(t, &wantRejectedResponse, rejectedResponse)
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:1072]: 💡 We're missing a test for a payment with `500+ GOAT` where the sender is KYC approved.

**Comment by Zagan202** [comment]
> 1. Inspect if transaction is a revised compliant and signed by the issuer(if not signed then have the issuer sign it).
> 
> I don't think this phrase is accurate, the word "revised" means something that was changed/edited. In this case the incoming transaction is already compliant with the regulated assets authorization strategy described in SEP-8, meaning no revision took place. Does that make sense?

You're totally right let me update that.

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go [hunk 4]
```diff
@@ -564,3 +763,311 @@ func TestTxApproveHandlerTxApprove(t *testing.T) {
 	}
 	assert.Equal(t, &wantRejectedResponse, rejectedResponse)
 }
+
+func TestTxApproveHandlerCheckIfCompliantTransaction(t *testing.T) {
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:767]: This function has over 300 lines of code now, we could split it into smaller chunks so it's easier to debug in case it starts failing. What about this division?

1. a test method for the approved case(s), non-KYC related
2. a test method for rejected cases
3. a test method test for KYC required cases

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 5]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+		return nil, errors.Wrap(err, "signing transaction")
+	}
+
+	// Encode revised transaction for response.
+	txe, err := tx.Base64()
+	if err != nil {
+		return nil, errors.Wrap(err, "encoding revised transaction")
+	}
+
+	return NewSuccessTxApprovalResponse(txe, "Transaction is compliant and signed by the issuer. Ready to submit!"), err
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/tx_approve.go:315]: 💡  It's not necessarily "ready to submit", right? This may or may not still need the sender's approval. We could just make this sentence simpler with
```suggestion
	return NewSuccessTxApprovalResponse(txe, "Transaction is compliant and signed by the issuer."), err
```

**Review by marcelosalloum — COMMENTED** [review]
I added a few suggestions below ⬇️ . The most important ones are around the `expectedOperations` + `compareCompliantTransactionOperations` strategy, and about adding another test case for KYC-required payments.

File: services/regulated-assets-approval-server/internal/serve/api_tx_approve_test.go [hunk 6]
```diff
@@ -537,6 +537,179 @@ func TestAPI_RevisedIntegration(t *testing.T) {
...
+	op5, ok := tx.Operations()[4].(*txnbuild.AllowTrust)
+	require.True(t, ok)
+	assert.Equal(t, op5.Trustor, senderAccKP.Address())
+	assert.Equal(t, op5.Type.GetCode(), assetGOAT.GetCode())
+	require.False(t, op5.Authorize)
+
+	// Check if issuer's signature is present.
+	txHash, err := tx.Hash(handler.networkPassphrase)
+	require.NoError(t, err)
+	err = handler.issuerKP.Verify(txHash[:], tx.Signatures()[0].Signature)
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/api_tx_approve_test.go:709]: `Verify` requires a byte array whereas `tx.Hash` returns specifically `[32]byte` so we convert it to a a `[]byte`.
Similar logic is seen [here](https://github.com/stellar/go/blob/master/txnbuild/transaction.go#L1262)

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 7]
```diff
@@ -68,6 +70,77 @@ func convertThresholdToReadableString(threshold int64) (string, error) {
 	return fmt.Sprintf("%.2f", res), nil
 }
 
+// compareCompliantTransactionOperations compares two txnbuild.Operation arrays and returns true if they're equivalent.
+// Hardcoded to check operation in order of what is expected from a composed compliant transaction.
+func compareCompliantTransactionOperations(expectedOperations, incomingOperations []txnbuild.Operation) (bool, error) {
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:75]: Does this read a bit better?
```suggestion
// compareCompliantTransactionOperations compares the incoming transactions operations and their order with the expected transactions and their order from a compliant transaction.
func compareCompliantTransactionOperations(expectedOperations, incomingOperations []txnbuild.Operation) (bool, error) {
```

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 8]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
 		return NewRejectedTxApprovalResponse("The source account is invalid."), nil
 	}
 
-	if len(tx.Operations()) != 1 {
-		return NewRejectedTxApprovalResponse("Please submit a transaction with exactly one operation of type payment."), nil
+	return nil, tx
+}
+
+// checkSequenceNum checks if transaction's sequence number is equivalent to source account's sequence number+1.
+func (h txApproveHandler) checkSequenceNum(ctx context.Context, tx *txnbuild.Transaction, acc horizon.Account) (resp *txApprovalResponse, err error) {
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:203]: Thats a good point I can change that.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 9]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+	}
+	if tx.SourceAccount().Sequence != accountSequence+1 {
+		log.Ctx(ctx).Errorf(`invalid transaction sequence number tx.SourceAccount().Sequence: %d, accountSequence+1: %d`, tx.SourceAccount().Sequence, accountSequence+1)
+		return NewRejectedTxApprovalResponse("Invalid transaction sequence number."), nil
+	}
+
+	return nil, nil
+}
+
+// checkIfCompliantTransaction inspects incoming transaction is compliant by wallets preemptively or by the server(according to the transaction-composition section of SEP-008).
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:216]: This is the rare case where we send a revised transaction back to the server via `POST /tx-approve`, although it would be a poor integration on the client side there's nothing in the SEP that states that the wallet cant do this.

Perhaps theres a scenario where the wallet/client has a retry approval attempt on an unsent `tx` and sends a `tx` back to the server just to ensure we're still compliant(or to make sure if the user needs KYC since at the time when they attempt to submit).

Regardless, I was describing that we're not constrained by having an incoming `tx`s that are only preemptively built by wallets.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 10]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+	if !ok {
+		log.Ctx(ctx).Error(`third operation is not of type payment`)
+		return NewRejectedTxApprovalResponse("There is one or more unauthorized operations in the provided transaction."), nil
+	}
+	paymentSource := paymentOp.SourceAccount
+	if paymentSource == "" {
+		paymentSource = tx.SourceAccount().AccountID
+	}
+	if paymentSource == h.issuerKP.Address() {
+		log.Ctx(ctx).Error(`transaction contains one or more operations where sourceAccount is issuer account`)
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:234]: I can see how that can read as misleading.
How does this sound?
```suggestion
		log.Ctx(ctx).Error(`transaction 's payment source account is equivalent to the issuer account`)
```

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 11]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
 		return NewRejectedTxApprovalResponse("There is one or more unauthorized operations in the provided transaction."), nil
 	}
+	// AllowTrustOp only stores the AssetCode (4- or 12-char string),but does not store the issuer.
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:237]: I feel like there is still debate if it is a bug or a feature(see this [comment describing it as a feature](https://stellarfoundation.slack.com/archives/C9H5JTW1J/p1622038808052400?thread_ts=1622000846.052100&cid=C9H5JTW1J) and this [comment describing it as a bug](https://stellarfoundation.slack.com/archives/C02B04RMK/p1621999545136100?thread_ts=1621984426.132900&cid=C02B04RMK)). I'm just describing the behavior without labeling as such until we reach an agreement.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 12]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+	// Since the issuer won't be in the encoded XDR we need to create a CreditAsset we should expect(which is one without an issuer).
+	// TODO: Update SEP-008 spec to replace AllowTrustOps(now deprecated) with SetTrustLineFlags operation.
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:239]: issue made [here](https://github.com/stellar/stellar-protocol/issues/951).
I'll remove the TODOs but I prefer to keep the other comments.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 13]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+	expectedAssetType := txnbuild.CreditAsset{
+		Code:   h.assetCode,
+		Issuer: "",
+	}
 
-	return nil, tx
+	// Compare incoming transactions with expected transactions.
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go:245]: Good catch!

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 2 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:793]: Actually you brought up a good point I overlooked, I dont need to mock `AccountDetail` response for the issuerKP or the receiverAccKP.

I'll remove those in a subsequent commit.

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 3 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:1072]: I initially thought that would be redundant to our other success case but you're right it would make the coverage complete.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: I made some major updates to the is function and how its called [here](5e0e4e21d833dfc0b20dc839dd993177d8ee5084)
👀 Could you take a look and see if we're on the same page?

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: Addressed [here](5e0e4e21d833dfc0b20dc839dd993177d8ee5084)

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: There's a change in progress that is going to inject the issuer address from the source account address https://github.com/stellar/go/pull/3636.
I believe we should move forward with what we have now and update the code and comments after its been merged.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: Removed [here](https://github.com/stellar/go/pull/3636)

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: Addressed [here](https://github.com/stellar/go/pull/3636)

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go [hunk 14]
```diff
@@ -123,16 +196,123 @@ func (h txApproveHandler) validateInput(ctx context.Context, in txApproveRequest
...
+			SourceAccount: h.issuerKP.Address(),
+		},
+		&txnbuild.AllowTrust{
+			Trustor:       paymentSource,
+			Type:          expectedAssetType,
+			Authorize:     false,
+			SourceAccount: h.issuerKP.Address(),
+		},
+	}
+	ok, err = compareCompliantTransactionOperations(expectedOperations, tx.Operations())
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: 👀 PTAL at what I did [here](5e0e4e21d833dfc0b20dc839dd993177d8ee5084).
I think this is a bit more maintainable, but I would love to know your thoughts.

File: services/regulated-assets-approval-server/internal/serve/tx_approve.go (hunk 5 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve.go]: Updated [here](868cf4c1cb338b7a88b6f9e201580f068f61ed4c)

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 1 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go]: Good catch thanks for pointing that out.
61952d65509e7371ac555d50765ec43c40e98adb

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 2 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go]: I removed horizon mocks that were not required [here](61952d65509e7371ac555d50765ec43c40e98adb)

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 3 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:1063]: Ive added this test case [here](61952d65509e7371ac555d50765ec43c40e98adb).

File: services/regulated-assets-approval-server/internal/serve/tx_approve_test.go (hunk 4 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go:774]: I'm totally down for this split!
Let me go ahead and refactor that.

Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/tx_approve_test.go]: I split them up here in this commit df2a89e.
A note about the naming schema for the test functions; I followed the `TestMyFunction_WhatIsBeingTested` schema for these split tests based on [`Uber Go Style Guide`'s](https://github.com/uber-go/guide/blob/master/style.md#function-names) suggestion

**Comment by marcelosalloum** [comment]
Closed in favor of https://github.com/stellar/go/pull/3652.
